### ISSUE TITLE
EdgeEdgeIntersection Bug

### ISSUE THREAD

**Issue by DiiiiiZ** [andywiecko/BurstTriangulator#173]
Hello ! :)

I discovered an edge case bug inside the EdgeEdgeIntersection function. In the case where two edges are not intersecting but are both on the same line, it cause the EdgeEdgeIntersection function to return true.

What happens ? 
The cross product should be returning zero, but instead returns 1E-13 (for example) due to floating point error (also happening for double) when using utils.neq and utils.sign (might also be an issue for other functions).

Here is the data i had:
```c#
double2 A0 = new double2(-703d, 290.42d);
double2 A1 = new double2(-701.15d, 288.57d);
double2 B0 = new double2(-193.43d, -219.15d);
double2 B1 = new double2(-186.58d, -226d);
``` 
With A0 and A1 corresponding to an edge, and B0 and B1 to another edge.

I tried some stuff to fix this like:
```c#
public readonly bool neq(double v, double w) => math.abs(v - w) > 1E-6d;
public readonly double sign(double a) => math.sign(neq(a, 0) ? a : 0);
``` 
The issue was gone but some other deeper issues appeared, so i'm sure there is a better solution.

Thank you :)

**Comment by andywiecko** [comment]
Hi @DiiiiiZ,
Thank you for reporting this issue! I'm surprised this bug came out just now! :)

The edge vs. edge intersection check is very fast but does not account for collinear edges. I believe this is not a problem for most uses of this utility in the package. However, your reported bug proves that this routine fails for extreme cases of non-intersecting collinear lines.

I first drew your input to show the extremity of your case. Here are the lines at true scale:

![image](https://github.com/user-attachments/assets/c2436942-ecff-4f59-9c1e-4357a42e46f9)

To solve the issue, I changed the CCW definition in my tests:

```diff
+ int CCW(double2 a, double2 b, double2 c) => (c.y - a.y) * (b.x - a.x) > (b.y - a.y) * (c.x - a.x) ? 1 : 0;
- int CCW(double2 a, double2 b, double2 c) => math.sign(Cross(b - a, b - c));
```

This elliminates the problem with float/double precision. I wrote a draft test to verify if this resolves the issue:

```csharp
private static bool EdgeEdgeIntersection(double2 a0, double2 a1, double2 b0, double2 b1) =>
    CCW(a0, a1, b0) != CCW(a0, a1, b1) && CCW(b0, b1, a0) != CCW(b0, b1, a1);
private static int CCW(double2 a, double2 b, double2 c) => (c.y - a.y) * (b.x - a.x) > (b.y - a.y) * (c.x - a.x) ? 1 : 0;

private static readonly TestCaseData[] edegeVsEdgeTestData = new TestCaseData[]
{
    new(
        math.double2(0, 0),
        math.double2(1, 0),
        math.double2(2, 0),
        math.double2(2, 1)
    ) { TestName = "Case 1 (ortogonal without intersection)", ExpectedResult = false},
    new(
        math.double2(-1, 0),
        math.double2(1, 0),
        math.double2(0, -1),
        math.double2(0, 1)
    ) { TestName = "Case 2 (ortogonal with intersection)", ExpectedResult = true},
    new(
        math.double2(0, 0),
        math.double2(1, 0),
        math.double2(2, 0),
        math.double2(3, 0)
    ) { TestName = "Case 3 (collinear without intersection)", ExpectedResult = false},
    new(
        math.double2(0, 0),
        math.double2(2, 0),
        math.double2(1, 0),
        math.double2(3, 0)
    ) { TestName = "Case 4 (collinear with intersection)", ExpectedResult = true},
    new(
        math.double2(-703d, 290.42d),
        math.double2(-701.15d, 288.57d),
        math.double2(-193.43d, -219.15d),
        math.double2(-186.58d, -226d)
    ) { TestName = "Case 5 (gh issue #173)", ExpectedResult = false},
}.SelectMany(i =>
{
    var args = i.Arguments;
    return new[]{
        new TestCaseData(args[0], args[1], args[2], args[3]) { TestName = i.TestName + ": abcd", ExpectedResult = i.ExpectedResult},
        new TestCaseData(args[0], args[1], args[3], args[2]) { TestName = i.TestName + ": abdc", ExpectedResult = i.ExpectedResult},
        new TestCaseData(args[1], args[0], args[2], args[3]) { TestName = i.TestName + ": bacd", ExpectedResult = i.ExpectedResult},
        new TestCaseData(args[1], args[0], args[3], args[2]) { TestName = i.TestName + ": badc", ExpectedResult = i.ExpectedResult},
    };
}
).ToArray();

[Test, TestCaseSource(nameof(edegeVsEdgeTestData))]
public bool EdegeVsEdgeTest(double2 a0, double2 a1, double2 b0, double2 b1) => EdgeEdgeIntersection(a0, a1, b0, b1);
```

In the previous implementation, your test data did not pass, but now it's green :)

![image](https://github.com/user-attachments/assets/88950b90-63d3-4599-a0c0-0783c7ded22c)

Intersecting collinear test cases still fail, but this should not be a problem, since we check for collinearity at input validation step. I know how to fix this as well, but due to performance reasons, I will first investigate if this is even necessary.

---

I am currently preparing the new release `v3.1` and your issue will definitely be included (with tests)! I just need a little time to prepare this properly (with some CCW math cleanup), but I think the release will be scheduled for next week.

Please let me know if this resolves the issue.

Best,
Andrzej

**Comment by andywiecko** [comment]
PS I did not check if this CCW implementation breaks something else, so during test please add additional CCW without replacing the previous one.

**Comment by DiiiiiZ** [comment]
Hey Andrzej ! Thank you for your quick answer :)

First i tried to replace the CCW implementation with your new one, and it did break at other places.
In all places we check CCW against utils.Zero(), like in "PointLineSegmentIntersection" or "IsConvexQuadrilateral", it breaks with the new implementation.

So i did as you suggested and used the new "EdgeEdgeIntersectionCCW" implementation only for the "EdgeEdgeIntersection" function, and it worked !

In conclusion, there is definitely something to fix here, but its in the right path :)
That said, I think it could be a lot easier by avoiding float point errors by using int. I believe i saw that it was work in progress on your side ? If so, do you have any updates about it ?

(Also, it might be really nice to directly log positions and edges positions in addition to indexes inside the EdgeValidations functions to facilitate debugging)

Thank you !

**Comment by andywiecko** [comment]
> So i did as you suggested and used the new "EdgeEdgeIntersectionCCW" implementation only for the "EdgeEdgeIntersection" function, and it worked !

Great, I'm glad it helped!

> That said, I think it could be a lot easier by avoiding float point errors by using int. I believe i saw that it was work in progress on your side ? If so, do you have any updates about it ?

Yes, I see that the CCW part can be simplified and refactored a bit. This is probably a good time to add tests for internal utilities. The current development status is as follows:

1. I want to release a new version next week (`v3.1`). This will include:
   - [X] Managed input (merged)
   - [X] Unsafe triangulator triangulate api (merged)
   - [X] Constrained halfedges in output data (merged)
   - [ ] unsafe triangulator refine and plant seeds api (draft ready)
   - [ ] Fix edge-edge issue + CCW refactor (draft ready)
 2. In the next release (`v3.2`), I plan to add very exciting features, I hope to release this next month (most of the features are already draft, last step is to implement the tests)!
    - [ ] Support for `int2` (draft ready)
    - [ ] Support for `fp64` (float fixed point in Q31.32 format), which should be interesting for you if you are looking for an integer like implementation. (draft ready)
    - [ ] Suport for `Vector2` (via reinterpreted, draft ready)
    - [ ] Dynamic point insertion/deletion via Unsafe triangulator api (initially, support will be added just for bulk operations in the mesh, however, this feature might be published in `v3.3` as well)

> (Also, it might be really nice to directly log positions and edges positions in addition to indexes inside the EdgeValidations functions to facilitate debugging)

Good idea! I've just reported this on the project board: https://github.com/users/andywiecko/projects/1/views/1?pane=issue&itemId=72267533

---

I will let you know when the fix for your issue is released.

If you have any additional questions or suggestions, let me know!

Best,
Andrzej

**Comment by DiiiiiZ** [comment]
Alright, that's good news !

Yes let me know when thoses fixes are released :)

Thank you for your hard work and your availability !

**Comment by andywiecko** [comment]
Hi @DiiiiiZ
I have great news! 

The new release, `v3.1`, is already published. You can find the full release notes [**here**](https://github.com/andywiecko/BurstTriangulator/releases/tag/v3.1.0)). It is already available on OpenUPM.

This should resolve your issue. Thank you for your contribution to the project! 
If you have any additional questions or suggestions, let me know.

Best, 
Andrzej
